File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    struct compare {
        bool operator()(ListNode* a, ListNode* b) {
            return a->val > b->val;  // min-heap: smallest value comes first
        }
    };

    ListNode* mergeKLists(vector<ListNode*>& lists) {
        priority_queue<ListNode*, vector<ListNode*>, compare> minHeap;

        // Step 1: Push head of each non-null list into the heap
        for (ListNode* node : lists) {
            if (node != nullptr) {
                minHeap.push(node);
            }
        }

        // Step 2: Dummy node for result list
        ListNode* dummy = new ListNode(-1);
        ListNode* curr = dummy;

        // Step 3: Build the merged list
        while (!minHeap.empty()) {
            ListNode* smallest = minHeap.top();
            minHeap.pop();

            curr->next = smallest;
            curr = curr->next;

            if (smallest->next != nullptr) {
                minHeap.push(smallest->next);
            }
        }

        return dummy->next;
    }
};

File: 0023-merge-k-sorted-lists/0023-merge-k-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
//  time : O(k * k * n)
// space : O(1)
class Solution {
    public ListNode mergeKLists(ListNode[] lists) {
       ListNode res = null;
       for(ListNode curr : lists){
        res = merge(res, curr);
       }
       return res;
    }
    //   TC: O(n+m) ~ k*n =>  n = number of ele in each list. 
    // n+ 2n+ 3n +....n*k
 //SC: O(1)
    ListNode merge(ListNode h1, ListNode h2){
       // 1. Create a dummy node to track the head of the resultant list
        //1. crete dummy node to track head of the resultant LL
        ListNode dummy = new ListNode(-1);
        ListNode curr = dummy; 

        //2. traverse from both ll and (add, iterate) in the smaller one 
        while(h1!=null && h2!=null){
            if(h1.val <= h2.val){
            curr.next=h1;
              h1= h1.next;
            }
            else{
                curr.next = h2;
                h2=h2.next;
            }

             curr=curr.next;
        }
        //traverse through either remaining non empty list
        if(h1!=null){
            curr.next = h1;
        }
        else{
            curr.next=h2;
        }

        return dummy.next;
    }
}